Split getElement{,sArray}Attribute into internal/bindings versions.

alice opened the pull request:
#### 4af3d6f6ef684cd3330948538aa51ed0dccef313
<pre>
Split getElement{,sArray}Attribute into internal/bindings versions.
<a href="https://bugs.webkit.org/show_bug.cgi?id=287538">https://bugs.webkit.org/show_bug.cgi?id=287538</a>

Reviewed by Ryosuke Niwa.

After referenceTarget is implemented, the internal version may return elements
within shadow roots, while the bindings version will ensure there is no leak of
elements out of shadow roots.

This change also switches to using Element::getElementForAttributeInternal()
and Element::getElementsArrayForAttributeInternal() instead of
TreeScope::getElementById() for resolving form association based on the
`form` attribute, label association using the `for` attribute, and list
association using the `list` attribute.

* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::addRelation):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::elementsFromAttribute const):
* Source/WebCore/bindings/js/JSElementCustom.cpp:
(WebCore::getElementsArrayAttribute):
* Source/WebCore/bindings/js/JSElementInternalsCustom.cpp:
(WebCore::getElementsArrayAttribute):
* Source/WebCore/bindings/scripts/CodeGenerator.pm:
(GetterExpression):
* Source/WebCore/bindings/scripts/test/JS/JSTestObj.cpp:
(WebCore::jsTestObj_reflectedElementAttrGetter):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::elementForAttributeInternal const):
(WebCore::Element::getElementAttributeForBindings const):
(WebCore::Element::elementsArrayForAttributeInternal const):
(WebCore::Element::getElementsArrayAttributeForBindings const):
(WebCore::Element::getElementAttribute const): Deleted.
(WebCore::Element::getElementsArrayAttribute const): Deleted.
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/ElementInternals.cpp:
(WebCore::ElementInternals::getElementAttributeForBindings const):
(WebCore::ElementInternals::getElementsArrayAttributeForBindings const):
(WebCore::ElementInternals::getElementAttribute const): Deleted.
(WebCore::ElementInternals::getElementsArrayAttribute const): Deleted.
* Source/WebCore/dom/ElementInternals.h:
* Source/WebCore/html/FormListedElement.cpp:
(WebCore::findAssociatedForm):
* Source/WebCore/html/HTMLButtonElement.cpp:
(WebCore::HTMLButtonElement::commandForElement const):
* Source/WebCore/html/HTMLFormControlElement.cpp:
(WebCore::HTMLFormControlElement::popoverTargetElement const):
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::dataList const):
* Source/WebCore/html/HTMLLabelElement.cpp:
(WebCore::elementForAttributeIfLabelable):
(WebCore::HTMLLabelElement::control const):
(WebCore::firstElementWithIdIfLabelable): Deleted.

Canonical link: <a href="https://commits.webkit.org/290360@main">https://commits.webkit.org/290360@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/971f104209512597ee944f31ab6d444af6c73d3a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89713 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/9240 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/44589 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/94706 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40480 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/91765 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/9629 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/17519 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/69098 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26722 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92714 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7393 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81404 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49464 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/7114 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/35914 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/39795 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77451 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/36813 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/96533 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16895 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/12390 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77962 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/17151 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/77226 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77294 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19091 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21723 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/20293 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/10065 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16908 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/22224 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16649 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/20100 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/18431 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -5261,10 +5261,10 @@ bool AXObjectCache::addRelation(Element& origin, const QualifiedName& attribute)
         return false;
     if (Element::isElementReflectionAttribute(Ref { m_document->settings() }, attribute)) {
-        if (auto reflectedElement = origin.getElementAttribute(attribute))
+        if (auto reflectedElement = origin.getElementForAttributeInternal(attribute))
```
If we're adding an internal variant of it, we should name it `elementForAttributeInternal` instead.
We typically don't use "get" prefixes unless it's to match web exposed API.

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2398,13 +2410,16 @@ void Element::setElementAttribute(const QualifiedName& attributeName, Element* e
+        auto it = map->find(attributeName);
+        if (it != map->end()) {
+            hasExplicitlySetElements = true;
+            elements = compactMap(it->value, [&](auto& weakElement) -> std::optional<Ref<Element>> {
```
Why we can't keep the early exit here?

alice replied:
Thanks for catching that - we can here, we just can't in the later PR, and I forgot to fix it here. Will fix.

Commit:
Split getElement{,sArray}Attribute into internal/bindings versions.
https://bugs.webkit.org/show_bug.cgi?id=287538

Reviewed by Ryosuke Niwa.

After referenceTarget is implemented, the internal version may return elements
within shadow roots, while the bindings version will ensure there is no leak of
elements out of shadow roots.

This change also switches to using Element::getElementForAttributeInternal()
and Element::getElementsArrayForAttributeInternal() instead of
TreeScope::getElementById() for resolving form association based on the
`form` attribute, label association using the `for` attribute, and list
association using the `list` attribute.

* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::addRelation):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::elementsFromAttribute const):
* Source/WebCore/bindings/js/JSElementCustom.cpp:
(WebCore::getElementsArrayAttribute):
* Source/WebCore/bindings/js/JSElementInternalsCustom.cpp:
(WebCore::getElementsArrayAttribute):
* Source/WebCore/bindings/scripts/CodeGenerator.pm:
(GetterExpression):
* Source/WebCore/bindings/scripts/test/JS/JSTestObj.cpp:
(WebCore::jsTestObj_reflectedElementAttrGetter):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::elementForAttributeInternal const):
(WebCore::Element::getElementAttributeForBindings const):
(WebCore::Element::elementsArrayForAttributeInternal const):
(WebCore::Element::getElementsArrayAttributeForBindings const):
(WebCore::Element::getElementAttribute const): Deleted.
(WebCore::Element::getElementsArrayAttribute const): Deleted.
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/ElementInternals.cpp:
(WebCore::ElementInternals::getElementAttributeForBindings const):
(WebCore::ElementInternals::getElementsArrayAttributeForBindings const):
(WebCore::ElementInternals::getElementAttribute const): Deleted.
(WebCore::ElementInternals::getElementsArrayAttribute const): Deleted.
* Source/WebCore/dom/ElementInternals.h:
* Source/WebCore/html/FormListedElement.cpp:
(WebCore::findAssociatedForm):
* Source/WebCore/html/HTMLButtonElement.cpp:
(WebCore::HTMLButtonElement::commandForElement const):
* Source/WebCore/html/HTMLFormControlElement.cpp:
(WebCore::HTMLFormControlElement::popoverTargetElement const):
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::dataList const):
* Source/WebCore/html/HTMLLabelElement.cpp:
(WebCore::elementForAttributeIfLabelable):
(WebCore::HTMLLabelElement::control const):
(WebCore::firstElementWithIdIfLabelable): Deleted.

Canonical link: https://commits.webkit.org/290360@main
